winpr/libwinpr/clipboard/posix.c

```diff
@@ -655,21 +686,62 @@ static void* convert_filedescriptors_to_file_list(wClipboard* clipboard, UINT32
 		const FILEDESCRIPTORW* cur = &descriptors[x];
 		size_t curLen = _wcsnlen(cur->cFileName, ARRAYSIZE(cur->cFileName));
 		char* curName = NULL;
+		char* stop_at = NULL;
+		char* previous_at = NULL;
 		rc = ConvertFromUnicode(CP_UTF8, 0, cur->cFileName, (int)curLen, &curName, 0, NULL, NULL);
 
-		rc = _snprintf(&dst[pos], alloc - pos, "%s%s/%s%s", lineprefix,
-		               clipboard->delegate.basePath, curName, lineending);
-		free(curName);
+		rc = _snprintf(&dst[pos], alloc - pos, "%s%s/", lineprefix, clipboard->delegate.basePath);
 
 		if (rc < 0)
 		{
 			free(dst);
 			return NULL;
 		}
+		pos += (size_t)rc;
+
+		previous_at = curName;
+		while ((stop_at = stop_at_special_chars(previous_at)) != NULL)
+		{
+			char* tmp = strndup(previous_at, stop_at - previous_at);
+			if (!tmp)
+			{
+				free(dst);
+				free(curName);
+				return NULL;
+			}
```

Comment:
This whole escaping thing gets really hard to read and reason about... I wonder if it can be improved, but I don't have any concrete ideas about it. Tough, as this does not seem to be like a very performance-sensitive piece of code, I guess it might be more simple to just read character by character and escape them as necessary, instead of using all those fancy C string functions.

Or maybe it will be better to first construct a string with printf() in a buffer – that makes it clear what the string actually contains. Then that buffer can be passed over to some escape_uri() function which will make a quick pass to check whether there *is* something to escape. If not then we're on a fast path, return the buffer unmodified. If there is, well then process it.